Restructure of Docker command execution and large build refactor

agjohnson opened the pull request:
**Note**: this is ongoing, review for design decisions for now

I was not happy with the pattern used to execute commands in containers. The original flow for executing tasks in Docker was:

* Get build/project/version state from API
* Serialize JSON
* Execute django admin command in docker, pass JSON into forked docker command via STDIN, docker command pipes to django admin command
* Django admin command deserialized JSON
* Execute build task from admin command, passing in project/version data
* Build task executes
* Serialize response, output STDOUT from django admin
* docker command returns STDOUT to original task

This was deficient is several ways:

* It required our full stack in a container - this was a very large container that needed to be re-rolled every deploy
* It is very fragile, requiring several layers of piping STDIN/STDOUT through commands.
* Breakage was difficult to track down when it did happen

**So**, this aimed to tackle that and turned into a larger refactor, as we could be more intelligent about how we were executing commands. My first inclination was to use proper IPC to make this less fragile, but @ericholscher brought up the point that we would treat the container as a dumb container as well. The goals of this PR will be:

- [x] Wrap all command execution in stateful `Environment` objects, so all forked commands are tracked.
- [x] A `LocalEnvironment` provides local command execution
- [x] Remove the need for `shell=True` command execution
- [x] Remove passing around ``results`` object in task -- `Environment` object tracks whether a command failed and has state on output/etc
- [x] Make docker image a lightweight shell, with only python dependencies
- [x] Docker commands execute through `exec` mechanism instead of a single command
- [x] Handle all docker commands through `docker-py`
- [x] Docker commands have no stdin/stdout, so output will be combined. This seems like a good time to drop the weird pattern of splitting stderr/stdout. I have thoughts on how to finally get combined stdout/stderr with formatting on stderr, but it's a surprisingly hard problem to solve.

Side effects of this refactor so far include:
- [x] Reduce duplicated code and build complexity greatly by removing magical PDF builds. Refactoring the commands executed, we were duplicating efforts to create a secondary PDF build. This build is ofter confusing for users, and shouldn't exist.
- [x] Put all commands for PDF/ePub builds in main build, fail task but don't halt on these errors. This will take some handling of exceptions during the build process and rewrap them if they are critical
- [x] Build output page is now a list of commands and output, not broken up stdout/stderr per build stage
- [x] No more build "steps" -- I'm still contemplating whether this would be useful, or if it would be useful in a differently applied method than it is now.
- [x] Task `update_docs` is now a subclassed `Task` object, allowing us to track state for the task. This could also be a separate class, but seemed more applicable as a full `Task` object with build/version/project state.

Finally, some of the design decisions here:
* `Environments` (`LocalEnvironment`/`DockerEnvironment`) are used to track build state, instead of passing around results and other states through method calls on project build
* `Commands` no longer require running in a `with` context
* `Environments` create these command objects, tracking and executing them
* `Environment` wrap all calls with exception handling. If an unhandled exception bubbles up to the Environment it is reported.
* A special exception is used to report errors to the user via builds page, `BuildEnvironmentError`. This will add error state to the command on the build page

Thoughts on where to take this next, and I think this work sets us up for an easy transition:
* Per-command output tracking -- this might make sense to include here, as the output is changing significantly already
* Additional environments for execution -- these will just be wrappers around executing commands, but could be anything (spin up ec2 instance, build, destroy/etc)

ericholscher commented on readthedocs/core/management/commands/update_docker.py:
```diff
@@ -20,4 +20,6 @@ def handle(self, *args, **options):
                 p = tasks.make_api_project(project_data)
                 log.info("Building %s" % p)
-                tasks.update_docs(pk=p.pk, docker=True)
+                # TODO just move this to update_api.py
```
Do we still need this file?

agjohnson replied:
We should decide what management commands are worth sticking around. Either way, build vs docker build should just be a parameter to the build comand.

ericholscher replied:
Agreed. We should figure out what makes sense to keep around, and document the useful ones.

agjohnson replied:
Also, maybe we consider implementing them as admin functions instead. Cases like build docs for a version from the admin are more useful for rtd.com, but allowing non ops folks to trigger some of the management functionality from the admin might be a win

ericholscher replied:
Yea, they should all be super light wrappers around util methods, so we could expose them both places.

ericholscher replied:
Made #1526 for this.

ericholscher commented on readthedocs/doc_builder/backends/mkdocs.py:
```diff
@@ -127,21 +126,18 @@ def append_conf(self, **kwargs):
-        results = run(build_command, shell=True)
-        return results
+        # TODO does this have the self.project?
+        # TODO --theme fails on the JSON builder
```
That is probably what's breaking mkdocs search.

agjohnson commented:
Outstanding issues that I am finishing up on:
- [ ] On docker command timeout, the container goes away as expected, but we should tell the user their build timed out on build output page
- [ ] Ensure container is cleaned up on exceptions, or be smarter about creating containers. We don't have access to the `--rm` option for `run` through the docker api for some reason, so the container sticks around on faulty builds
- [ ] Build output page is still awful, will address separately so this PR can continue

ericholscher commented on readthedocs/core/management/commands/update_api.py:
```diff
@@ -14,10 +14,15 @@ class Command(BaseCommand):
                 p = tasks.make_api_project(project_data)
                 log.info("Building %s" % p)
-                tasks.update_docs(pk=p.pk)
+                update_docs = tasks.UpdateDocsTask()
```
Is there a reason we don't just put this in the tasks.py so it's importable like previously?

ericholscher commented on readthedocs/doc_builder/backends/mkdocs.py:
```diff
@@ -154,3 +148,13 @@ class MkdocsJSON(BaseMkdocs):
     builder = 'json'
     build_dir = '_build/json'
+
+    def build(self, **kwargs):
```
Does this need to be subclassed? It seems like the same logic as the class above?

ericholscher replied:
Oh I see, it kills the theme passing in.

ericholscher replied:
Would be better to do a theme = True/False instead of repeating the logic.

agjohnson replied:
good call, i kept what was there already, but it's a good spot for clean up

agjohnson commented:
Okay, to wrap this up, I've added some documentation and cleaned up the last remaining issues that I had with the code. I've push up a new repo with the docker file required to start with this build env: https://github.com/rtfd/readthedocs-docker-images

It would be worth throwing some more projects at this system for integration testing. I've been through many iterations of testing on this all, but would be good to revisit initial setup issues/etc.

I will follow this up with another PR that makes the UX on output not suck. It currently is a first pass and is quite noisey.

gregmuellegger commented on docs/development/buildenvironments.rst:
```diff
@@ -0,0 +1,68 @@
+==================
+
+Read the Docs uses container virtualization to encapsulate documentation build
+processes. Each build spins up a new virtual machinge using our base image,
```
type: "machinge" -> "machine"

gregmuellegger commented on docs/development/buildenvironments.rst:
```diff
@@ -0,0 +1,68 @@
+
+There are several settings used to configure usage of virtual machines:
+
+    DOCKER_ENABLED
```
Intending the definition list here will make it live inside a blockquote. The definition list should start in column 0.

gregmuellegger commented on readthedocs/doc_builder/environments.py:
```diff
@@ -207,105 +76,471 @@ def run(self, cmd_input=None, combine_output=False):
     def get_command(self):
         '''Flatten command'''
-        if isinstance(self.command, list):
+        if hasattr(self.command, '__iter__'):
```
This works in Python 2, as strings have no `__iter__` method. In Python 3, it will return `True` and therefore will insert a space between each char of the command. One solution to that is to add `not isinstance(self.command, str)`.

gregmuellegger commented on readthedocs/rtd_tests/tests/test_privacy.py:
```diff
@@ -15,24 +15,13 @@
-
-        self.tester = User(username='tester')
-        self.tester.set_password('test')
-        self.tester.save()
```
That is undoing the refactor from #1532
See https://github.com/rtfd/readthedocs.org/pull/1532/files#diff-27d2585b4d0b054a45164644132a4e5eL17

agjohnson replied:
good catch, this got dropped while rebasing, i'll add it back today

ericholscher commented:
:shipit:
